AsyncDisplayKit/Layout/ASLayoutablePrivate.h

```diff
@@ -60,9 +59,202 @@ extern void ASLayoutableClearCurrentContext();
  */
 @property (nonatomic, assign) BOOL isFinalLayoutable;
 
+@end
+
+
+#pragma mark - ASLayoutOptionsForwarding
 
 /**
- *  The class that holds all of the layoutOptions set on an ASLayoutable.
+ *  Both an ASDisplayNode and an ASLayoutSpec conform to ASLayoutable. There are several properties
+ *  in ASLayoutable that are used when a node or spec is used in a layout spec.
+ *  These properties are provided for convenience, as they are forwards to the node or spec's
+ *  properties. Instead of duplicating the property forwarding in both classes, we
+ *  create a define that allows us to easily implement the forwards in one place.
+ *
+ *  If you create a custom layout spec, we recommend this stragety if you decide to extend
+ *  ASDisplayNode and ASLayoutSpec to provide convenience properties for any options that your
+ *  layoutSpec may require.
  */
-@property (nonatomic, strong, readonly) ASLayoutOptions *layoutOptions;
-@end
+
+#define ASEnvironmentLayoutOptionsForwarding \
+- (void)propagateUpLayoutOptionsState\
+{\
+  id<ASEnvironment> parent = [self parent];\
+  if ([parent supportsUpwardPropagation]) {\
+    ASEnvironmentStatePropagateUp(parent, _environmentState.layoutOptionsState);\
+  }\
+}\
+\
+- (CGFloat)spacingAfter\
+{\
+  ASDN::MutexLocker l(_propertyLock);\
+  return _environmentState.layoutOptionsState.spacingAfter;\
+}\
+\
+- (void)setSpacingAfter:(CGFloat)spacingAfter\
+{\
+  _propertyLock.lock();\
+  _environmentState.layoutOptionsState.spacingAfter = spacingAfter;\
```

Comment:
Fantastic, all of these methods are what I had hoped for in terms of a great balance between efficiency and clarity. I think this implementation will last a long time without any modification or optimization.

Indeed, the performance of these methods will greatly exceed that of the node bridged properties, especially because calling into the underlying Apple frameworks adds even more overhead on top of the MutexLockers in all of the setter methods.